src/std/make.ss

```diff
@@ -278,28 +315,50 @@
          (vector-set! deps@ target deps)))))
   (when (verbose>=? 3) (writeln [Step: 3 deps: (apply map list (map vector->list [id@ spec@ deps@]))]))
 
-  ;; 4. Create output directories
-  (create-directory* (settings-bindir settings))
-  (create-directory* (settings-libdir settings))
-  (when (settings-static settings) (create-directory* (path-expand "static" (settings-libdir settings))))
-  (when (verbose>=? 3) (writeln [Step: 4]))
-
-  ;; 5. Initialize the blocked and blocking tables and the build queue
-  ((cut vector-for-each <> deps@)
-   (lambda (target deps)
-     ((cut for-each <> deps)
-       (lambda (dep)
-         (let ((other-target (hash-get id-index% dep)))
-           (if other-target
-             (begin (hash-put! (vector-ensure-ref blocked-by@ target make-hash-table) other-target #t)
-                    (hash-put! (vector-ensure-ref blocking@ other-target make-hash-table) target #t))
-             (dependency-timestamp dep))))))) ; NB: this populates the timestamp% table as a side-effect
-  ((cut vector-for-each <> spec@)
-   (lambda (target _) (unless (vector-ref blocked-by@ target) (pqueue-push! ready target))))
-  (when (verbose>=? 3) (writeln [Step: 5 first-target: (vector-ref spec@ (pqueue-peek ready))]))
+  ;; 4. Update the build-deps cache, topologically sorted.
+  (def (initialize-walk) ;; 4. Initialize a walk on the dependencies
+    (vector-clear! blocked-by@)
+    (vector-clear! blocking@)
+    (assert! (pqueue-empty? ready-fg))
+    (assert! (pqueue-empty? ready-bg))
+    (set! build-list '())
+
+    ((cut vector-for-each <> deps@)
+     (lambda (target deps)
+       ((cut for-each <> deps)
+        (lambda (dep)
+          (let ((other-target (hash-get id-index% dep)))
+            (when other-target
+              (hash-put! (vector-ensure-ref blocked-by@ target make-hash-table) other-target #t)
+              (hash-put! (vector-ensure-ref blocking@ other-target make-hash-table) target #t)))))))
+    ((cut vector-for-each <> spec@)
+     (lambda (target _) (unless (vector-ref blocked-by@ target) (pqueue-push! ready-fg target)))))
+
+  (def (mark-built target . _)
+    (push! target build-list)
+    (alet (blocking (vector-ref blocking@ target))
+      ((cut hash-for-each <> blocking)
+       (lambda (unblocked _)
+         (def blocked-by (vector-ref blocked-by@ unblocked))
+         (hash-remove! blocked-by target)
+         (when (hash-empty? blocked-by)
+           (pqueue-push! ready-fg unblocked))))))
+
+  (initialize-walk)
+  (perform-plan/threads
+   ready-fg ready-bg
+   perform: void
+   announce: void
+   on-success: mark-built on-failure: (cut error "impossibru" <> <>)
```

Comment:
dunno how impossible that error is -- shouldn't we print a more meaningful error message?